File: Selenium/Activities/activity21.java
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

public class activity21 {
    public static void main(String[] args) {
        System.setProperty(FirefoxDriver.SystemProperty.BROWSER_LOGFILE, "/dev/null");
        WebDriverManager.firefoxdriver().setup();
        WebDriver driver = new FirefoxDriver();
        driver.get("https://www.training-support.net/selenium/tab-opener");
        System.out.println("The title of page is: " + driver.getTitle());
        System.out.println("Current tab: " + driver.getWindowHandle());
        //Find the button to open a new tab and click it.
        driver.findElement(By.id("launcher")).click();
        //Wait for the new tab to open.
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        wait.until(ExpectedConditions.numberOfWindowsToBe(2));
        //Print all the handles.
        System.out.println("Currently open windows: " + driver.getWindowHandles());
        //Switch to the newly opened tab, print it's title and heading.
        for(String handle : driver.getWindowHandles()) {
            driver.switchTo().window(handle);
        }
        wait.until(ExpectedConditions.elementToBeClickable(By.id("actionButton")));
        // Print the handle of the current tab
        System.out.println("Current tab: " + driver.getWindowHandle());
        // Print the title and heading of the new page
        System.out.println("Page title: " + driver.getTitle());
        String pageHeading = driver.findElement(By.className("content")).getText();
        System.out.println("Page Heading: " + pageHeading);
        // Find and click the button on page to open another tab
        driver.findElement(By.id("actionButton")).click();
        //Repeat the steps by clicking the button in the new tab page.
        wait.until(ExpectedConditions.numberOfWindowsToBe(3));
        //Print all the handles.
        System.out.println("Currently open windows: " + driver.getWindowHandles());
        //Switch to the newly opened tab, print it's title and heading.
        for(String handle : driver.getWindowHandles()) {
            driver.switchTo().window(handle);
        }
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("content")));
        // Print the handle of the current tab
        System.out.println("Current tab: " + driver.getWindowHandle());
        // Print the title and heading of the new page
        System.out.println("Page title: " + driver.getTitle());
        String pageHeading1 = driver.findElement(By.className("content")).getText();
        System.out.println("Page Heading: " + pageHeading1);
        driver.quit();

    }
}

File: Selenium/Activities/Activity21.py
from selenium import webdriver
from selenium.webdriver.firefox.service import Service as FirefoxService
from webdriver_manager.firefox import GeckoDriverManager
from selenium.webdriver.remote.webdriver import By
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


service = FirefoxService(GeckoDriverManager().install())

with webdriver.Firefox(service=service) as driver:
    driver.get("https://www.training-support.net/selenium/tab-opener")
    print("Page title is: ", driver.title)
    print("Current window handle: ", driver.current_window_handle)
    #Find the button to open a new tab and click it.
    driver.find_element(By.ID,"launcher").click()
    #Wait for the new tab to open.
    wait = WebDriverWait(driver,10)
    wait.until(EC.number_of_windows_to_be(2))
    #Print all the handles.
    print("All window handles available are: ", driver.window_handles)
    #Switch to the newly opened tab, print it's title and heading.
    driver.switch_to.window(driver.window_handles[1])
    print("The current handle is: ", driver.current_window_handle)
    print("The heading of the page is: ", driver.title)
    #Repeat the steps by clicking the button in the new tab page.
    wait.until(EC.visibility_of_element_located((By.ID,"actionButton")))
    driver.find_element(By.ID,"actionButton").click()
    wait.until(EC.number_of_windows_to_be(3))
    #Print all the handles.
    print("All window handles available are: ", driver.window_handles)
    #Switch to the newly opened tab, print it's title and heading.
    driver.switch_to.window(driver.window_handles[2])
    print("The current handle is: ", driver.current_window_handle)
    print("The heading of the page is: ", driver.title)
